src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java

```diff
@@ -0,0 +1,123 @@
+
+import static org.hamcrest.Matchers.*;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class AnimalWebControllerTest {
+
+	private final String path = "/web/animal";
+	private final String viewPackage = "animal/";
+	private List<Animal> animals;
+	private MockMvc mockMvc;
+	@MockBean
+	private AnimalService animalService;
+	@Autowired
+	private WebApplicationContext webApplicationContext;
+
+	@Before
+	public void setUp() {
+		mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext).build();
+
+		Animal lion = new Animal(1L, "Lion", "King of Africa");
+		Animal tiger = new Animal(2L, "Tiger", "The Bengali predator");
+		Animal elephant = new Animal(3L, "Elephant", "Biggest terrestrial mammal");
+		Animal dauphin = new Animal(4L, "Dauphin", "Smartest aqua mammal");
+		Animal mountainLion = new Animal(5L, "Lion", "Mountain Lion, King of Atlas");
+		animals = Arrays.asList(lion, tiger, elephant, dauphin, mountainLion);
+	}
+
+	@Test
+	public void getAnimalsTest() throws Exception {
+		// Happy path
+		when(animalService.getAnimals()).thenReturn(animals);
+
+		mockMvc.perform(get(path + "/list"))
+				.andExpect(status().isOk())
+				.andExpect(view().name(viewPackage + "animal_list"))
+				.andExpect(model().attribute("animals", hasSize(5)));
+
+		verify(animalService, times(1)).getAnimals();
+		verifyNoMoreInteractions(animalService);
+	}
+
+	@Test
+	public void getAnimalByExistingId() throws Exception {
+		// Happy path
+		when(animalService.getAnimalById(3L)).thenReturn(animals.get(2));
+
+		mockMvc.perform(get(path + "/list/3"))
+				.andExpect(status().isOk())
+				.andExpect(view().name(viewPackage + "animal_details"))
+				.andExpect(model().attribute("animal", is(animals.get(2))));
+
+		verify(animalService, times(1)).getAnimalById(3L);
+		verifyNoMoreInteractions(animalService);
+	}
+
+	@Test
+	public void getAnimalByNonExistingId() throws Exception {
+		// Sad path
+		when(animalService.getAnimalById(6L)).thenThrow(AnimalNotFoundException.class);
+
+		mockMvc.perform(get(path + "/list/6"))
+				.andExpect(status().isNotFound());
+
+		verify(animalService, times(1)).getAnimalById(6L);
+		verifyNoMoreInteractions(animalService);
+	}
+
+	@Test
+	public void addAnimalTest() throws Exception {
+		// Happy path
+		mockMvc.perform(get(path + "/add"))
+				.andExpect(status().isOk())
+				.andExpect(model().attribute("animal", hasProperty("id", nullValue())))
+				.andExpect(view().name(viewPackage + "animal_form"));
+
+		verifyZeroInteractions(animalService);
+	}
+	// There is no Sad path for requesting the "/add"
+
+	@Test
+	public void saveNewAnimalTest() throws Exception {
```

Comment:
what about Sad path here? But I guess that based on my previous comment, the error handling will change a bit, so you can consider that (or you could try TDD approach for that :D)

Reply:
Requesting that path only redirect to the form and doesn't interact with the service. So there is no sad path fortunately.

Reply:
well then you can test that it will do the redirect, right? But as I said, that will probably change when you do the error handling inside the form